Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -398,6 +399,36 @@ - (void)writeAvatarToDisk:(UIImage *)avatar
     });
 }
 
+- (NSData *)processedImageDataForRawAvatar:(UIImage *)image
+{
+    NSUInteger kMaxAvatarBytes = 5 * 1000 * 1000;
+
+    if (image.size.width != kOWSProfileManager_MaxAvatarWidth
+        || image.size.height != kOWSProfileManager_MaxAvatarWidth) {
+        // To help ensure the user is being shown the same cropping of their avatar as
+        // everyone else will see, we want to be sure that the image was resized before this point.
+        OWSFail(@"Avatar image should have been resized before trying to upload");
+        image = [image resizedImageToFillPixelSize:CGSizeMake(kOWSProfileManager_MaxAvatarWidth,
+                                                       kOWSProfileManager_MaxAvatarWidth)];
+    }
+
+    NSData *_Nullable data;
+    for (NSUInteger attempts = 0; attempts < 5; attempts++) {
+        CGFloat quality = (CGFloat)0.95 - attempts * (CGFloat)0.1;
+        data = UIImageJPEGRepresentation(image, quality);
+        if (data.length <= kMaxAvatarBytes) {
+            return data;
+        } else {
+            // This for-loop is really just paranoia. Our avatar dimensions are so small that
+            // it's incredibly unlikely we wouldn't be able to fit our profile photo with even
+            // our highest quality.
```

Comment:
The stress test will be to use as input rgb noise, since I think that'll be worst-case for jPEG compression.  If even that compresses under 5mb, let's take out this loop since a) all clients should be robust to oversize avatars b) simplicity!